src/Network/Transport/TCP/Internal.hs

```diff
@@ -162,22 +186,47 @@ forkServer host port backlog reuseAddr terminationHandler requestHandler = do
       when reuseAddr $ N.setSocketOption sock N.ReuseAddr 1
       N.bindSocket sock (N.addrAddress addr)
       N.listen sock backlog
+
+      -- Close up and fill the synchonizing MVar.
+      let release :: ((N.Socket, N.SockAddr), MVar ()) -> IO ()
+          release ((sock, _), socketClosed) =
+            N.sClose sock `finally` putMVar socketClosed ()
+
+      -- Run the request handler.
+      let act :: (N.Socket, N.SockAddr) -> IO ()
+          act (sock, sockAddr) = do
+            socketClosed <- newEmptyMVar
+            void $ forkIO $ do
+              requestHandler (readMVar socketClosed) sock
+              `finally`
+              release ((sock, sockAddr), socketClosed)
+
+      let acceptRequest :: IO ()
+          acceptRequest = mask $ \restore -> do
+            -- Async exceptions are masked so that, if accept does give a
+            -- socket, we'll always deliver it to the handler before the
+            -- exception is raised.
+            -- If it's a Right handler then it will eventually be closed.
+            -- If it's a Left handler then we assume the handler itself will
+            -- close it.
+            (sock, sockAddr) <- N.accept sock
+            -- Looks like 'act' will never throw an exception, but to be
+            -- safe we'll close the socket if it does.
+            let handler :: SomeException -> IO ()
+                handler _ = N.sClose sock
+            catch (restore (act (sock, sockAddr))) handler
```

Comment:
Ok I'll make the change. I can't see how it could go wrong as is, but I'm also fairly certain that moving the restore into the forked thread would not introduce any problems.